R/get_ci.R

```diff
@@ -83,95 +167,297 @@ get_ci <- function(dist, interval, estimate, variance = NULL, denominator = NULL
-
-  # wrangle output
-  result <- ci %>%
-    data.frame() %>%
-    data.table::transpose() %>%
-    dplyr::rename(lower = 1, upper = 2)
+  q_upper <- 1 - (alpha / 2)
+  q_lower <- alpha / 2
+
+  result <- data.frame(x = x, y = y) %>%
+    dplyr::mutate(
+      lower = stats::qchisq(q_lower, df = 2 * x) / (2 * y),
+      upper = stats::qchisq(q_upper, df = 2 * (x + 1)) / (2 * y)
+    ) %>%
+    dplyr::select(.data$lower, .data$upper)
 
   return(result)
 }
 
 
-#' Replicate `stats::qpois` using `stats::qchisq`
+#' Calculate a gamma distribution confidence interval
+#'
+#' @description Derive an approximate gamma distribution confidence interval
+#' using [stats::qchisq()].
+#'
+#' @param interval A scalar between 0 and 1, indicating the width of the
+#' interval. For example, use `0.95` to calculate a 95% confidence interval.
+#' @param estimate A numeric vector containing the point estimates.
+#' @param weights A list of numeric vectors containing the weights.
+#' @param variance A numeric vector containing the variances.
+#' @param method A string to indicate which method to use to calculate the
+#' confidence interval. The default is `"tcz06"` which refers to the method
+#' proposed by Tiwari, Clegg, and Zou (2006). Use `"ff97"` for a more
+#' conservative confidence interval proposed by Fay and Feuer (1997).
 #'
-#' @description This function implements the Poisson distribution using the
-#' chi-squared distribution to allow for the continuous extension of integer
-#' variable (`estimate`). Read p.9385 of the [STDRATE Procedure](https://support.sas.com/documentation/onlinedoc/stat/151/stdrate.pdf)
-#' for more information.
+#' @details This function assumes that the values of the input arguments
+#' `estimate`, `weights`, and `variance` are aligned and that they are the same
+#' length.
 #'
-#' @param estimate A scalar, must be a whole number (e.g. number of events)
-#' @param denominator A scalar (e.g. population-time)
-#' @param q_low A scalar, indicating the lower limit
-#' @param q_high A scalar, indicating the upper limit
+#' This function is used to add confidence intervals in [get_ds_rt()].
 #'
-#' @return A numeric vector containing the lower and upper limits
+#' @return A data frame with columns `lower` and `upper` which contain the lower
+#' and upper limits of a confidence interval respectively.
+#'
+#' @references \href{http://documentation.sas.com/doc/en/pgmsascdc/9.4_3.4/statug/statug_stdrate_details.htm}{The STDRATE Procedure}
 #'
 #' @examples
 #' \dontrun{
-#' get_ci_pois(q_low = 0.025, q_high = 0.975, estimate = 30, denominator = 1000)
+#' # calculate a single confidence interval
+#' e_1 <- 0.015
+#' w_1 <- c(
+#'   3.7e-05, 2.2e-05, 2.0e-05, 1.2e-05, 2.3e-05,
+#'   7.0e-05, 3.8e-05, 6.0e-05, 1.7e-05, 2.6e-05
+#' )
+#' v_1 <- 1e-06
+#'
+#' get_ci_gamma(interval = 0.95, estimate = e_1, weights = list(w_1), variance = v_1)
+#' get_ci_gamma(interval = 0.9, estimate = e_1, weights = list(w_1), variance = v_1, method = "ff97")
```

Comment:
Suggest showing an example that shows the difference between the two methods, i.e., one example for the default "tcz06" and one example for "ff97", for the same width interval.

I'm also noticing that there seems to be no change between the lower limits of the confidence intervals produced by the two methods, which is consistent with how it is coded (Lines 415, 422, 432), however I just wanted to double check if this is how the intervals produced by the two methods should be, according to the methodology.

Reply:
Yes, that is correct, only the upper confidence interval limit changes between `ff97` and `tcz06`. See: http://documentation.sas.com/doc/en/pgmsascdc/9.4_3.4/statug/statug_stdrate_details14.htm#statug_stdrate002267 where it says (bold added)

> Tiwari, Clegg, and Zou (2006) propose a less conservative confidence interval for  with a different **upper** confidence limit